File: 43左旋转字符串/Solution.java
public class Solution {
    public String LeftRotateString(String str,int n) {
       if(n>=str.length())
           return str;
        char[] c=str.toCharArray();
        reverse(c,0,n-1);
        reverse(c,n,str.length()-1);
        reverse(c,0,str.length()-1);
        return new String(c);
    }
    private void reverse(char[] c,int i,int j){
        while (i<j)
            swap (c,i++,j--);
    }
    private void swap(char[] c,int i,int j){
        char t=c[i];
        c[i]=c[j];
        c[j]=t;
    }
}
// 方法与cpp方法相同

File: 43左旋转字符串/Solution.cpp
class Solution {
public:
    string LeftRotateString(string str, int n) {
        string result = str;
        int length = result.size();
        if(length < 0){
            return NULL;
        }
        if(0 <= n <= length){
            int pFirstBegin = 0, pFirstEnd = n - 1;
            int pSecondBegin = n, pSecondEnd = length - 1;
            ReverseString(result, pFirstBegin, pFirstEnd);
            ReverseString(result, pSecondBegin, pSecondEnd);
            ReverseString(result, pFirstBegin, pSecondEnd);
        }
        return result;
    }
private:
    void ReverseString(string &str, int begin, int end){
        while(begin < end){
            swap(str[begin++], str[end--]);
        }
    }
};
// 先将数组的前n个元素逆置，然后把后n个元素逆置，最后在全体逆置